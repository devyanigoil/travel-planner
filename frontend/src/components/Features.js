import React, { useState } from 'react';
import './Features.css';

const Features = () => {
  const [features, setFeatures] = useState([
    { id: 1, title: 'Feature 1', content: 'Description of Feature 1' },
    { id: 2, title: 'Feature 2', content: 'Description of Feature 2' },
    { id: 3, title: 'Feature 3', content: 'Description of Feature 3' },
  ]);

  return (
    <div className="features">
      <div className="left-div">
        <h1>Our Services</h1>
      </div>
      <div className="right-div">
        <div
          className="accordion custom-accordion border-0"
          id="featuresAccordion"
        >
          {features.map((feature) => (
            <div
              className="accordion-item custom-accordion-item"
              key={feature.id}
            >
              <div className="custom-accordion-header">
                <h2
                  className="accordion-header"
                  id={`heading-${feature.id}`}
                >
                  <button
                    className="accordion-button collapsed d-flex justify-content-between align-items-center bg-white shadow-none custom-accordion-button"
                    type="button"
                    data-bs-toggle="collapse"
                    data-bs-target={`#collapse-${feature.id}`}
                    aria-expanded="false"
                    aria-controls={`collapse-${feature.id}`}
                  >
                    <span>{feature.title}</span>
                    <span className="toggle-icon"></span>
                  </button>
                </h2>
              </div>

              <div
                id={`collapse-${feature.id}`}
                className="accordion-collapse collapse"
                aria-labelledby={`heading-${feature.id}`}
                data-bs-parent="#featuresAccordion"
              >
                <div className="accordion-body">{feature.content}</div>
              </div>
            </div>
          ))}
        </div>
      </div>
    </div>
  );
};

export default Features;
