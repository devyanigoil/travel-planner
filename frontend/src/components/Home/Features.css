.features {
    padding: 0px !important;
    margin: 0px !important;
    margin-top: 5rem !important;
    display: grid;
    grid-template-columns: 1fr; /* Change to single column for small screens */
    grid-gap: 2rem;
    background-color: #589ea8;
  }
  
  /* Adjust layout on larger screens */
  @media (min-width: 768px) {
    .features {
      grid-template-columns: 1fr 1fr; /* 2 columns on larger screens */
    }
  }
  
  /* Left Div Adjustments */
  .left-div {
    display: flex;
    justify-content: flex-start;
    align-items: flex-start;
    height: 100%;
    /* margin-left: 1.5rem; Adjust left margin */
    padding-top: 1.5rem; /* Space from top */
  }
  
  .left-div h1 {
    font-size: 2.5rem; /* Adjust font size */
    font-weight: bold;
    color: white;
    padding-left: 2rem; /* Space from left */
    padding-top: 1rem; /* Space from top */
    margin: 0;
  }
  
  /* Right Div Adjustments */
  .right-div {
    display: flex;
    justify-content: center;
    align-items: flex-start;
    padding-right: 2rem;
  }
  
  /* Accordion Customizations */
  .custom-accordion-item {
    border-top: 1px solid black !important;
    border-left: 0 !important;
    border-right: 0 !important;
    background-color: #589ea8 !important;
  }
  
  .custom-accordion-header .accordion-button {
    background-color: #589ea8 !important;
    color: black !important;
    border: none;
  }
  
  .custom-accordion-header .accordion-button:hover {
    background-color: #589ea8 !important;
    color: #000 !important;
  }
  
  .custom-accordion-button {
    font-size: 1.5rem !important;
    padding-bottom: 1.5rem !important;
    padding-top: 1.5rem !important;
    width: 40rem !important;
  }
  