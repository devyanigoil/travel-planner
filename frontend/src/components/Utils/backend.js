import axios from "axios";

export const CONNECTION_URL = `http://localhost:8081`;

const getFeaturesList = () => {
  return axios.get(CONNECTION_URL + `/getFeatures`);
};

const addUser = (formData)=>{
  return axios.post(CONNECTION_URL + '/addUser', formData);
}

const backend = {
  getFeaturesList,
  addUser
};

export default backend;
