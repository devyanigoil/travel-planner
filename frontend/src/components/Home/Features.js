import React, { useEffect, useState } from "react";
import "./Features.css";
import backend from "../Utils/backend";

const Features = () => {
  const [features, setFeatures] = useState([]);

  const getFeatureList = async () => {
    try {
      const response = await backend.getFeaturesList();
      const updatedFeatures = response.data[0].features.map(
        (feature, index) => ({
          ...feature,
          id: index + 1, // Add an ID (index-based or any unique identifier logic)
        })
      );
      setFeatures(updatedFeatures);
    } catch (error) {
      console.error("Something went wrong");
    }
  };

  useEffect(() => {
    getFeatureList();
  }, []);

  return (
    <div className="features">
      <div className="left-div">
        <h2>Our Services</h2>
      </div>
      <div className="right-div">
        <div
          className="accordion custom-accordion border-0"
          id="featuresAccordion"
        >
          {features.map((feature) => (
            <div
              className="accordion-item custom-accordion-item"
              key={feature.id}
            >
              <div className="custom-accordion-header">
                <h2 className="accordion-header" id={`heading-${feature.id}`} data-testid={`heading-${feature.id}`}>
                  <button
                    className="accordion-button collapsed d-flex justify-content-between align-items-center bg-white shadow-none custom-accordion-button"
                    type="button"
                    data-bs-toggle="collapse"
                    data-bs-target={`#collapse-${feature.id}`}
                    aria-expanded="false"
                    aria-controls={`collapse-${feature.id}`}
                  >
                    <span>{feature.feature}</span>
                    <span className="toggle-icon"></span>
                  </button>
                </h2>
              </div>

              <div
                id={`collapse-${feature.id}`}
                className="accordion-collapse collapse"
                aria-labelledby={`heading-${feature.id}`}
                data-bs-parent="#featuresAccordion"
                data-testid={`collapse-${feature.id}`}
              >
                <div className="accordion-body">{feature.description}</div>
              </div>
            </div>
          ))}
        </div>
      </div>
    </div>
  );
};

export default Features;
