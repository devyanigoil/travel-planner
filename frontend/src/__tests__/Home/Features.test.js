import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import Features from '../../components/Home/Features';
import backend from '../../components/Utils/backend';

// Mock the backend
jest.mock('../../components/Utils/backend', () => ({
  getFeaturesList: jest.fn()
}));

describe('Features Component', () => {
  // Sample mock data for features
  const mockFeatures = [{
    features: [
      {
        feature: 'Trip Planning',
        description: 'Comprehensive trip planning service',
      },
      {
        feature: 'Packing Assistance',
        description: 'Smart packing recommendations',
      }
    ]
  }];

  beforeEach(() => {
    // Clear all mocks before each test
    jest.clearAllMocks();
  });

  describe('Initial Rendering', () => {
    test('renders "Our Services" heading', async () => {
      // Mock successful API call
      backend.getFeaturesList.mockResolvedValue({ data: mockFeatures });

      render(<Features />);

      // Check for the section heading
      const headingElement = screen.getByText('Our Services');
      expect(headingElement).toBeInTheDocument();
    });

    test('fetches and renders features correctly', async () => {
      // Mock successful API call
      backend.getFeaturesList.mockResolvedValue({ data: mockFeatures });

      render(<Features />);

      // Wait for features to be rendered
      await waitFor(() => {
        const tripPlanningFeature = screen.getByText('Trip Planning');
        const packingAssistanceFeature = screen.getByText('Packing Assistance');
        
        expect(tripPlanningFeature).toBeInTheDocument();
        expect(packingAssistanceFeature).toBeInTheDocument();
      });
    });
  });

  describe('API Interaction', () => {
    test('calls getFeaturesList on component mount', async () => {
      // Mock successful API call
      backend.getFeaturesList.mockResolvedValue({ data: mockFeatures });

      render(<Features />);

      // Verify that getFeaturesList was called
      await waitFor(() => {
        expect(backend.getFeaturesList).toHaveBeenCalledTimes(1);
      });
    });

    test('handles API error gracefully', async () => {
      // Mock console.error to prevent error logging in test output
      const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});

      // Mock API error
      backend.getFeaturesList.mockRejectedValue(new Error('API Error'));

      render(<Features />);

      // Wait for error handling
      await waitFor(() => {
        expect(backend.getFeaturesList).toHaveBeenCalledTimes(1);
        expect(consoleErrorSpy).toHaveBeenCalledWith('Something went wrong');
      });

      // Restore console.error
      consoleErrorSpy.mockRestore();
    });
  });

  describe('Accordion Rendering', () => {
    test('renders accordion with correct structure', async () => {
      // Mock successful API call
      backend.getFeaturesList.mockResolvedValue({ data: mockFeatures });

      render(<Features />);

      // Wait for features to be rendered
      await waitFor(() => {
        const features = mockFeatures[0].features;
        
        features.forEach((feature, index) => {
          // Check feature button
          const featureButton = screen.getByText(feature.feature);
          expect(featureButton).toBeInTheDocument();
          
          // Check feature description in accordion body
          const featureDescription = screen.getByText(feature.description);
          expect(featureDescription).toBeInTheDocument();

          // Check unique IDs for accordion items
          const accordionHeader = screen.getByTestId(`heading-${index + 1}`);
          const accordionCollapse = screen.getByTestId(`collapse-${index + 1}`);
          
          expect(accordionHeader).toBeInTheDocument();
          expect(accordionCollapse).toBeInTheDocument();
        });
      });
    });
  });

  // Snapshot test
  test('matches snapshot', async () => {
    // Mock successful API call
    backend.getFeaturesList.mockResolvedValue({ data: mockFeatures });

    const { asFragment } = render(<Features />);

    // Wait for features to be rendered
    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot();
    });
  });
});