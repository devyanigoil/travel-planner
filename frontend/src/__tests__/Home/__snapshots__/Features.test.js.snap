// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Features Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="features"
  >
    <div
      class="left-div"
    >
      <h2>
        Our Services
      </h2>
    </div>
    <div
      class="right-div"
    >
      <div
        class="accordion custom-accordion border-0"
        id="featuresAccordion"
      />
    </div>
  </div>
</DocumentFragment>
`;
